### PR TITLE
[FLINK-27856][flink-kubernetes]solve the NPE error of no spec field in taskmanager pod template.

### PR DESCRIPTION
**Pull request by Grypse**
## What is the purpose of the change

This pull request fix the NPE error of TM pod template file without spec field.

## Brief change log
Add validation for TM pod template file spec field.

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change added tests and can be verified as follows:
  - Added test that validates TM pod template without spec field: `KubernetesUtilsTest.testLoadPodFromNoSpecTemplate`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes(fixed the JM crash on kubernetes when tm pod template file without spec field.)

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by Grypse** [comment]
@wangyang0918 Please help me review this PR.

**Comment by Grypse** [comment]
> * @flinkbot run azure

**Review by wangyang0918 — COMMENTED** [review]
Could we fix this issue in the `KubernetesUtils#loadPodFromTemplateFile`? Set an empty spec if not specified.

**Comment by Grypse** [comment]
> Could we fix this issue in the `KubernetesUtils#loadPodFromTemplateFile`? Set an empty spec if not specified.

This PR fixes the issue in KubernetesUtils#loadPodFromTemplateFile. When no spec field is specified, set an empty one.

Code:
`    public static FlinkPod loadPodFromTemplateFile(
            FlinkKubeClient kubeClient, File podTemplateFile, String mainContainerName) {
        final KubernetesPod pod = kubeClient.loadPodFromTemplateFile(podTemplateFile);
        final List<Container> otherContainers = new ArrayList<>();
        Container mainContainer = null;

        if (null != pod.getInternalResource().getSpec()) {
            for (Container container : pod.getInternalResource().getSpec().getContainers()) {
                if (mainContainerName.equals(container.getName())) {
                    mainContainer = container;
                } else {
                    otherContainers.add(container);
                }
            }
            pod.getInternalResource().getSpec().setContainers(otherContainers);
        } else {
            // Set empty spec for taskmanager pod template
            pod.getInternalResource().setSpec(new PodSpecBuilder().build());
        }

        if (mainContainer == null) {
            LOG.info(
                    "Could not find main container {} in pod template, using empty one to initialize.",
                    mainContainerName);
            mainContainer = new ContainerBuilder().build();
        }

        return new FlinkPod(pod.getInternalResource(), mainContainer);
    }`

**Comment by wangyang0918** [comment]
My bad. I will take closer look now.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/KubernetesUtils.java [hunk 1]
```diff
@@ -419,12 +420,18 @@ public static FlinkPod loadPodFromTemplateFile(
...
+            for (Container container : pod.getInternalResource().getSpec().getContainers()) {
+                if (mainContainerName.equals(container.getName())) {
+                    mainContainer = container;
+                } else {
+                    otherContainers.add(container);
+                }
             }
+            pod.getInternalResource().getSpec().setContainers(otherContainers);
+        } else {
+            // Set empty spec for taskmanager pod template
```
Comment (wangyang0918) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/KubernetesUtils.java:433]: This is not only for taskmanger pod template.

**Review by wangyang0918 — APPROVED** [review]
I have left a minor comment. After it is addressed I believe this PR is ready for merging.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/KubernetesUtils.java (hunk 1 above)
Comment (Grypse) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/KubernetesUtils.java:433]: Right, it's also called by jobmanager pod template, but when no spec field on template, otherContainers is empty, and the PodSpec containers initial is an empty `List<Containers>` ,So there is no need to set an empty otherContainers to pod Spec.
PodSpec containers property:
```
    @JsonProperty("containers")
    private List<Container> containers = new ArrayList();
```

Comment (Grypse) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/KubernetesUtils.java]: Fixed.

**Review by wangyang0918 — APPROVED** [review]
+1

**Comment by Grypse** [comment]
@flinkbot run azure

**Comment by wangyang0918** [comment]
Would you like to create a backport PR to release-1.15?

**Comment by Grypse** [comment]
> Would you like to create a backport PR to release-1.15?

My pleasure!
The backport PR has been created to release-1.15: https://github.com/apache/flink/pull/20558

**Pull request merged**